File: AoC2017/Day03.fs
namespace AoC2017

open Common
open System.Collections.Generic
open System.Linq 
open System.Numerics

type Day03 = 
    inherit Solution

    new (file:string) = { inherit Solution(file) }
    new (input : List<string>) = { inherit Solution(input) }

    override this.SolvePart1() = 
        let size = ceil ( sqrt (this.Input.First() |> float )) |> int
        let isEven = size % 2 = 0 
        let offset = pown size 2 - (this.Input.First() |> int) |> fun (o:int) -> if isEven then o / 2 else o / 2 - 1
        let distance = if isEven then size - offset else size - offset - 1
        distance.ToString()

    override this.SolvePart2() = "" 

File: AoC2021/Day03.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Common;
using Common.Extensions;

namespace AoC2021
{
    public class Day03 : Solution
    {
        private const char One = '1';
        private const char Zero = '0';
        public Day03(string file) : base(file) { }

        public override string SolvePart1()
        {
            var gamma = new StringBuilder();
            var epsilon = new StringBuilder();

            for (var idx = 0; idx < Input[0].Length; idx++)
            {
                var (most, least) = GetMostAndLeastCommonBit(Input.Select(i => i[idx]));
                gamma.Append(most);
                epsilon.Append(least);
            }

            return (gamma.ToDecimal() * epsilon.ToDecimal()).ToString();
        }

        public override string SolvePart2()
        {
            var oxygen = GetRating(Input, One);
            var carbon = GetRating(Input, Zero);
            return (oxygen * carbon).ToString();
        }

        private static long GetRating(List<string> bits, char criteria)
        {
            var idx = 0;

            while (bits.Count > 1)
            {
                var (most, least) = GetMostAndLeastCommonBit(bits.Select(i => i[idx]));
                bits = bits.Where(i => i[idx] == (criteria == One ? most : least)).ToList();
                idx++;
            }

            return Convert.ToInt64(bits.First(), 2);
        }

        private static (char most, char least) GetMostAndLeastCommonBit(IEnumerable<char> bits)
        {
            var zero = bits.Count(c => c == Zero);
            var one = bits.Count(c => c == One);
            var most = zero > one ? Zero : One;
            var least = most == Zero ? One : Zero;
            return (most, least);
        }
    }
}